---
title: "Genome Annotation"
subtitle: "Finding available annotation and retrieving SNP annotation"
author: "Michelle Perez-Cuevas"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Setup

```{r setup}
### Genome annotation: Find & get
### Alexis Norris
### Created: 2023-06-22
### Revised: 2023-07-12

### Install required packages (only need to do once per computer)
#BiocManager::install("AnnotationHub")
#BiocManager::install("GenomeInfoDb")
#BiocManager::install("biomaRt")
#BiocManager::install("rtracklayer")
#install.packages("sessioninfo")
#install.packages("tidyverse")

### Load packages
library(tidyverse)
```

# AnnotationHub

Not available for pig.

# biomaRt

Reference: <https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html>

## Load

```{r biomart_load}
### Load the package
library(biomaRt)
```

## Find what data is available

Here, they're called Pages. And you have attributes (columns) you can select from a given Page (a table)

```{r biomart_find}
### What versions are available
### Most of the time we want the newest
listEnsemblArchives()

### Find databases
listEnsembl()$biomart

### Choose version and database
### If "useast" mirror down, use "www" or "asia"
Sys.time()
bm_snps <- useEnsembl(
  biomart = "snps", 
  version = 109
)
Sys.time()

### Find species
searchDatasets(
  mart = bm_snps, 
  pattern = "Sscrofa11.1|pig"
)
Sys.time()
bm_snps <- useDataset(
  mart = bm_snps, 
  dataset = "sscrofa_snp"
)
Sys.time()

### Find which datasets/attributes are available
bm_snps_attributes <- listAttributes(bm_snps)  
head(bm_snps_attributes)
table(bm_snps_attributes$page)
```

## Retrieve desired data

```{r biomart_retrieve}
### Get the annotation for selected attributes
Sys.time()
anno_biomart <- getBM(
  attributes = c(
    "refsnp_id", 
    "refsnp_source",
    "refsnp_source_description",
    "chr_name", 
    "chrom_start", 
    "chrom_end", 
    "chrom_strand", 
    "allele",
    "allele_1",
    "minor_allele",
    "minor_allele_freq",
    "minor_allele_count",
    "consequence_type_tv"
  ),  
  mart = bm_snps
)
Sys.time()
```

# rtracklayer

Reference: <https://bioconductor.org/packages/devel/bioc/vignettes/rtracklayer/inst/doc/rtracklayer.pdf> (see Part 5)

## Load

```{r rtrack_load}
### Load the package
library(rtracklayer)

### Load the annotation database
Sys.time()
session <- browserSession("UCSC")
Sys.time()
```

## Find what data is available

Here, the data are Tables' Tracks from [UCSC Table Browser](https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=1646468690_nm3tICVA49bscHaGtvBWdqxRG6mu)

```{r rtrack_find}
### Look what species are available 
### And subset for latest version of pig
ucscGenomes(organism = FALSE)[["db"]][grep("susScr", ucscGenomes(organism = FALSE)[["db"]])]
genome(session) <- "susScr11"

### Find out what types of data are available
### Tracks
track_names <- trackNames(ucscTableQuery(session))
track_names

### Tables (for a given track)
ucscTables(
  genome = "susScr11", 
  track = "evaSnp4"
)
```
```{r rtrack_find_all, eval=FALSE}
### This triggers error when run in Rmd

### Tables (for all tracks)
rtrack_all <- lapply(track_names, function(track) {
  tables <- ucscTables(
    genome = "susScr11", 
    track = track
  )
  tibble(
    "Table" = tables
  )
}) %>%
  bind_rows(.id = "Track")
write_csv(
  rtrack_all,
  "rtracklayer_datasets_susScr11.csv"
)
```

## Retrieve desired data

```{r rtrack_retrieve}
### Get the annotation for selected Track-Table
Sys.time()
anno_rtracklayer <- getTable(ucscTableQuery(
  session, 
  track = "evaSnp4",
  table = "evaSnp4"
))
Sys.time()
```

# What format is the annotation from each source?

```{r compare}
### Look at structure
class(anno_biomart)
class(anno_rtracklayer)

### Look at structure
str(anno_biomart)
str(anno_rtracklayer)

### Look at top
head(anno_biomart)
head(anno_rtracklayer)
```

# Methods

```{r versions}
### Versions
sessioninfo::session_info()
```
